File: apps/macos/Kelpie/Handlers/ScrollHandler.swift
import WebKit

// swiftlint:disable line_length

/// Handles scroll, scroll2, scrollToTop, scrollToBottom, scrollToY.
///
/// Every scroll operation also syncs `window.__m3d.scrollX/Y` when the 3D
/// inspector is active so that exiting the inspector restores to the
/// currently-visible position rather than the enter-time snapshot.
struct ScrollHandler {
    let context: HandlerContext

    /// JS fragment that syncs the 3D inspector's stored scroll state with the
    /// live `window.scrollX/Y`. A no-op when the inspector isn't active.
    private static let sync3DStateJS = """
    if (window.__m3d) {
        window.__m3d.scrollX = window.scrollX;
        window.__m3d.scrollY = window.scrollY;
    }
    """

    func register(on router: Router) {
        router.register("scroll") { body in await scroll(body) }
        router.register("scroll2") { body in await scroll2(body) }
        router.register("scroll-to-top") { _ in await scrollTo(top: true) }
        router.register("scroll-to-bottom") { _ in await scrollTo(top: false) }
        router.register("scroll-to-y") { body in await scrollToY(body) }
    }

    @MainActor
    private func scroll(_ body: [String: Any]) async -> [String: Any] {
        let dx = body["deltaX"] as? Double ?? 0
        let dy = body["deltaY"] as? Double ?? 0
        let js = """
        (function() {
            window.scrollBy(\(dx), \(dy));
            \(Self.sync3DStateJS)
            return {scrollX: window.scrollX, scrollY: window.scrollY};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func scroll2(_ body: [String: Any]) async -> [String: Any] {
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let position = body["position"] as? String ?? "center"
        let maxScrolls = body["maxScrolls"] as? Int ?? 10
        let js = """
        (function() {
            var el = document.querySelector('\(selector.replacingOccurrences(of: "'", with: "\\'"))');
            if (!el) return null;
            el.scrollIntoView({block: '\(position)', behavior: 'smooth'});
            \(Self.sync3DStateJS)
            var rect = el.getBoundingClientRect();
            return {element: {tag: el.tagName.toLowerCase(), visible: rect.top >= 0 && rect.bottom <= window.innerHeight, rect: {x: rect.x, y: rect.y, width: rect.width, height: rect.height}}, scrollsPerformed: 1, viewport: {width: window.innerWidth, height: window.innerHeight}};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            if result.isEmpty { return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)") }
            await context.showTouchIndicatorForElement(selector)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func scrollTo(top: Bool) async -> [String: Any] {
        let targetY = top ? "0" : "document.documentElement.scrollHeight"
        let js = """
        (function() {
            window.scrollTo(0, \(targetY));
            \(Self.sync3DStateJS)
            return {scrollX: window.scrollX, scrollY: window.scrollY};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func scrollToY(_ body: [String: Any]) async -> [String: Any] {
        guard let y = body["y"] as? Double else {
            return errorResponse(code: "MISSING_PARAM", message: "y is required (pixel offset)")
        }
        let x = body["x"] as? Double ?? 0
        let js = """
        (function() {
            window.scrollTo(\(x), \(y));
            \(Self.sync3DStateJS)
            return {
                scrollX: window.scrollX,
                scrollY: window.scrollY,
                maxScrollY: Math.max(0, document.documentElement.scrollHeight - window.innerHeight)
            };
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/ScrollHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class ScrollHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("scroll") { scroll(it) }
        router.register("scroll2") { scroll2(it) }
        router.register("scroll-to-top") { scrollTo(true) }
        router.register("scroll-to-bottom") { scrollTo(false) }
        router.register("scroll-to-y") { scrollToY(it) }
    }

    private suspend fun scroll(body: Map<String, Any?>): Map<String, Any?> {
        val dx = (body["deltaX"] as? Number)?.toDouble() ?: 0.0
        val dy = (body["deltaY"] as? Number)?.toDouble() ?: 0.0
        val js = "window.scrollBy($dx, $dy); ({scrollX: window.scrollX, scrollY: window.scrollY})"
        return try {
            successResponse(ctx.evaluateJSReturningJSON(js))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun scroll2(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val position = body["position"] as? String ?: "center"
        val safe = selector.replace("'", "\\'")
        val js =
            "(function(){var el=document.querySelector('$safe');" +
                "if(!el)return null;el.scrollIntoView({block:'$position',behavior:'smooth'});" +
                "var r=el.getBoundingClientRect();" +
                "return{element:{tag:el.tagName.toLowerCase()," +
                "visible:r.top>=0&&r.bottom<=window.innerHeight," +
                "rect:{x:r.x,y:r.y,width:r.width,height:r.height}}," +
                "scrollsPerformed:1,viewport:{width:window.innerWidth,height:window.innerHeight}};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result.isEmpty()) {
                errorResponse("ELEMENT_NOT_FOUND", "Element not found: $selector")
            } else {
                ctx.showTouchIndicatorForElement(selector)
                successResponse(result)
            }
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun scrollTo(top: Boolean): Map<String, Any?> {
        val js =
            if (top) {
                "window.scrollTo(0, 0); ({scrollY: 0})"
            } else {
                "window.scrollTo(0, document.documentElement.scrollHeight); ({scrollY: window.scrollY})"
            }
        return try {
            successResponse(ctx.evaluateJSReturningJSON(js))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun scrollToY(body: Map<String, Any?>): Map<String, Any?> {
        val y =
            (body["y"] as? Number)?.toDouble()
                ?: return errorResponse("MISSING_PARAM", "y is required (pixel offset)")
        val x = (body["x"] as? Number)?.toDouble() ?: 0.0
        val js = "(function(){window.scrollTo($x,$y);return{scrollX:window.scrollX,scrollY:window.scrollY,maxScrollY:Math.max(0,document.documentElement.scrollHeight-window.innerHeight)};})()"
        return try {
            successResponse(ctx.evaluateJSReturningJSON(js))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }
}
